sycl/include/sycl/ext/oneapi/experimental/graph.hpp

```diff
@@ -350,13 +384,60 @@ class command_graph : public detail::modifiable_command_graph {
 template <>
 class command_graph<graph_state::executable>
     : public detail::executable_command_graph {
-
 protected:
   friend command_graph<graph_state::executable>
   detail::modifiable_command_graph::finalize(const sycl::property_list &) const;
   using detail::executable_command_graph::executable_command_graph;
 };
 
+namespace detail {
+class __SYCL_EXPORT dynamic_parameter_base {
+public:
+  dynamic_parameter_base(
+      sycl::ext::oneapi::experimental::command_graph<graph_state::modifiable>
+          Graph);
+
+  void register_with_node(int ArgIndex, handler &CGH);
+
+protected:
+  void updateValue(void *NewValue, size_t Size);
+
+  void updateAccessor(sycl::detail::AccessorBaseHost *Acc);
+  std::shared_ptr<dynamic_parameter_impl> impl;
+};
+} // namespace detail
+
+template <typename ValueT>
+class dynamic_parameter : public detail::dynamic_parameter_base {
+  static constexpr bool IsAccessor =
+      std::is_base_of_v<sycl::detail::AccessorBaseHost, ValueT>;
+  static constexpr sycl::detail::kernel_param_kind_t ParamType =
+      IsAccessor ? sycl::detail::kernel_param_kind_t::kind_accessor
+                 : std::is_pointer_v<ValueT>
+                       ? sycl::detail::kernel_param_kind_t::kind_pointer
+                       : sycl::detail::kernel_param_kind_t::kind_std_layout;
+
+public:
+  dynamic_parameter(experimental::command_graph<graph_state::modifiable> Graph)
+      : detail::dynamic_parameter_base(Graph), MValue() {}
+
+  dynamic_parameter(ValueT InitialValue,
+                    experimental::command_graph<graph_state::modifiable> Graph)
+      : detail::dynamic_parameter_base(Graph), MValue(InitialValue) {}
```

Comment:
This constructor isn't in the current spec draft, and I know the advantage is that a user doesn't need to pass the template type to the constructor.
e.g.
```
exp_exp::dynamic_parameter(V, G);
````
rather than
```
exp_exp::dynamic_parameter<decltype(V)> (G);
```
We should decide to add this to the spec or not.

Reply:
I just pushed this to the spec PR, but can revert if there are objections to it. It might even make more sense to make it the only constructor, otherwise we probably need to define the behaviour of a `dynamic_parameter` which has never had it's value set by `update()`